feat: Block editor assets endpoint

dcalhoun opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Related: 

- CMM-221-linear-issue
- 177481-ghe-Automattic/wpcom
- https://github.com/wordpress-mobile/GutenbergKit/pull/109
- https://github.com/wordpress-mobile/WordPress-iOS/pull/24170
- https://github.com/wordpress-mobile/WordPress-Android/pull/21773

## Proposed changes:
Close CMM-221. Add a new REST API endpoint for retrieving the resources utilized in the block editor, to enable creating a block editor that supports third-party block scripts and styles outside of the WP Admin—e.g., within a mobile app WebView. 

Currently, we intend to use this endpoint for Simple sites, but plan to use it for Atomic and self-hosted, Jetpack-connected sites in the future. 

My perception is that this endpoint addition is unlikely to introduce regressions for Simple, Atomic, or self-hosted Jetpack-connected sites. I have not observed any regressions during my testing. 

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
pbArwn-763-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No, not of which I am aware. 

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

1. Create a WPCOM Simple site. 
2. Apply these Jetpack changes (see [comment](https://github.com/Automattic/jetpack/pull/42314#issuecomment-2707566476)) and 177481-ghe-Automattic/wpcom to your sandbox. 
3. Route your development machine traffic for `public-api.wordpress.com` to your sandbox via your `/etc/hosts` file.
5. Make request to the endpoint via the [API console](https://developer.wordpress.com/docs/api/console/) for a WPCOM Simple site. 
6. Verify the 200 response includes expected styles and scripts used by the block editor—e.g., `jetpack-connection-js`, `jetpack-blocks-editor-js`. 
    ```json
    {
        "styles": "<string_of_link_tags>",
        "scripts": "<string_of_script_tags>",
    }
    ```

dcalhoun committed:
feat: Add block editor assets endpoint

The REST API endpoint returns all the styles and scripts to be loaded
for the post editor. This is useful when recreating the `post-new.php`
editor environment in a different context--e.g., a WebView within a
mobile app. Access to the assets allows loading plugin-provided blocks.

dcalhoun committed:
feat: Match WP REST API editor assets paths

In additional to WordPress.com API URLs, we should match WP REST API
URLs as well. This is used for Atomic and self-hosted sites.

dcalhoun committed:
refactor: Relocate block editor assets endpoint to Jetpack

We plan to utilize this endpoint for all site types, so it makes sense
to place it within Jetpack now.

dcalhoun committed:
changelog

dcalhoun committed:
test: Add initial block editor assets endpoint tests

dcalhoun committed:
style: Address lint warnings

dcalhoun committed:
test: Attempt fix for CI-only test failure

The test fails in the "PHP tests: PHP 8.1 WP latest with wpcomsh" CI
test run with the following:

```
fopen(/tmp/wordpress-latest/src/wp-content/mu-plugins/wpcomsh/jetpack_vendor/automattic/jetpack-mu-wpcom/src/build/block-inserter-modifications/block-inserter-modifications.asset.php): Failed to open stream: No such file or directory
```

dcalhoun committed:
test: Attempt fix for CI-only test failure (2)

The test fails in the "PHP tests: PHP 8.1 WP latest with wpcomsh" CI
test run with the following:

```
fopen(/tmp/wordpress-latest/src/wp-content/mu-plugins/wpcomsh/jetpack_vendor/automattic/jetpack-mu-wpcom/src/build/block-inserter-modifications/block-inserter-modifications.asset.php): Failed to open stream: No such file or directory
```

dcalhoun committed:
test: Fix missing build asset errors

Not all assets are built for the CI environment, so we need to dequeue
those assets in the test to avoid errors.

dcalhoun committed:
test: Refactor test to align with existing test practices

dcalhoun committed:
feat: Remove manual loading of WP Admin resources

This is unneeded at this time, as the originally targeted Jetpack
resource (`jetpack-connect`) is no longer conditionally registered.

dcalhoun committed:
docs: Update documentation for Jetpack

dcalhoun committed:
test: Explore simplified assets queue

Limit the enqueued assets to avoid missing built asset files in the test
environment.

dcalhoun commented on projects/plugins/jetpack/.phan/baseline.php:
```diff
@@ -122,6 +122,7 @@
         '_inc/lib/core-api/class.jetpack-core-api-widgets-endpoints.php' => ['PhanTypeMismatchReturn'],
         '_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-admin-menu.php' => ['PhanPluginMixedKeyNoKey', 'PhanPluginSimplifyExpressionBool'],
         '_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-ai.php' => ['PhanPluginMixedKeyNoKey', 'PhanTypeMismatchArgument'],
+        '_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-block-editor-assets.php' => ['PhanPluginMixedKeyNoKey'],
```
Added so that the API endpoint structure might match existing endpoints in the code base.

dcalhoun committed:
style: Disable lint warning preventing schema alignment

The `schema` key was added for alignment with existing API endpoints.

enejb commented on projects/plugins/jetpack/tests/php/core-api/wpcom-endpoints/WPCOM_REST_API_V2_Endpoint_Block_Editor_Assets_Test.php:
```diff
@@ -0,0 +1,133 @@
+	 * Test that get_items returns the expected structure.
+	 */
+	public function test_get_items() {
+		wp_set_current_user( self::factory()->user->create( array( 'role' => 'editor' ) ) );
```
the lowest role that we would expect here to work should be contributor. 
More of a nit: but maybe we should check that one?

dcalhoun committed:
test: Verify editor assets endpoint with lowest permission level

The `contributor` role is the lowest possible role for accessing this
endpoint, so it is worth testing with this role.

enejb reviewed (dismissed):
nice work! 
 This worked as expected for me.

enejb commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-block-editor-assets.php:
```diff
@@ -0,0 +1,202 @@
+
+declare( strict_types = 1 );
+
+if ( ! class_exists( 'WPCOM_REST_API_V2_Endpoint_Block_Editor_Assets_Controller' ) ) {
```
Should this also be updated to match the class? I am not sure this is needed since this file should only be loaded once.

dcalhoun replied:
Good catch. I removed this in https://github.com/Automattic/jetpack/pull/42314/commits/2a796b3c4cee81229992f244e7c40901eeb8228c.

dcalhoun committed:
refactor: Remove unnecessary, broken duplicate definition conditional

The class name referenced was an outdated name. The check is also
unnecessary as this file should only be loaded once.
